SecureComms: Add support daemonConfig (#2065)
confidential-containers/cloud-api-adaptor: src/cloud-api-adaptor/docs/SecureComms.md

Here, we describe how to set up the Secure Communications (`Secure Comms` for short) feature of the `cloud-api-adaptor` (`Adaptor`) and `agent-protocol-forwarder` (`Forwarder`).

The `Secure Comms` feature establishes an SSH channel by the `Adaptor`, as an SSH client, and the `Forwarder`, as an SSH server. All control communications then use this SSH channel between the `Adaptor` and `Forwarder`. This feature is an alternative to using `TLS encryption` feature between the `Adaptor` and `Forwarder`. 

> Both `Secure Comms` and `TLS encryption` support a method of delivering secrets via cloud API userdata during VM creation. Delivering the secrets via cloud API userdata allows an attacker in the cloud API to gain access and modify any communication between all clusters and their respective Peer Pods on that cloud. For this reason, `Secure Comms` also supports a more secure mode that include the delivery of secrets via `Trustee`. Such secrets are then delivered to the Peer Pod after a an attestation phase (See diagram below). Delivering the secrets via `Trustee` protects the Peer Pod from an offender located at the cloud API and is the recommended method for Confidential Containers. We name this option `Trustee Mode` while the option of using `Secure Comms` with userdata is named `NoTrustee Mode`.

Reviewer: ```suggestion
Both `Secure Comms` and `TLS encryption` supports a method of delivering keys to encrypt the control-plane traffic between the `Adaptor` and the `Forwarder` via user-data during pod VM creation. Delivering the keys via the user-data may allow a malicious entity on the cloud provider side to gain access to the keys and modify any communication between the WNs and PPs. For this reason, `Secure Comms` also supports an alternate mode where the keys are delivered to the PP via `Trustee`. The keys are delivered to the PP after an attestation phase (See diagram below). Delivering the keys via `Trustee` protects the PP from a malicious entity on the cloud provider side. 
We name this mode of operation as `Trustee Mode`. The option of delivering keys via user-data is named  as `NoTrustee Mode`.
```

<p align="center">
  <img width="500"  src="SecureComms_TLS.png">
</p>

To deliver the secrets via `Trustee`, `Secure Comms` under `Trustee Mode` uses a two-phase approach:
- Attestation Phase: In this phase, an SSH channel is created to enable the Peer Pod to attest against an attestation service. The attestation traffic is tunneled from the Peer Pod to the CAA and from there sent to a KBS as configured at the CAA. Once attestation is achieved, the Forwarder obtains Peer Pod keys from the KBS. The Peer Pod keys are dynamically added to the KBS by the CAA prior to starting the Peer Pod.
- Kubernetes Phase: In this phase, an SSH channel is created to enable the runtime to communicate with the Kata Agent running in the Peer Pod. This phase uses the keys obtained during the Attestation Phase, ensuring that the communication is secured.

Additional tunnels can be established via the SSH channel by configuring the Adaptor and Forwarder as described below.

## Secure Comms Architecture

> `Secure Comms` uses a Go SSH Lib to secure the communication between the cluster Worker Nodes and Peer Pods. The Worker Nodes are assumed to be protected by a firewall and can be kept unreachable from outside of the cluster of Worker Nodes. To communicate with Peer Pods, an SSH channel is created from the `Adaptor` of the Worker Node allocated to run the Peer Pod and the Peer Pod VM `Forwarder`. This SSH channel can then be used for all communication between the Peer Pod and the cluster and any back end services running at the cluster or elsewhere.

Reviewer: ```suggestion
`Secure Comms` uses a Go SSH Lib to secure the communication between the cluster Worker Nodes (WNs) and Peer Pods (PPs). To communicate with the PPs, an SSH channel is created between the `cloud-api-adaptor` in the WN and the `agent-protocol-forwarder` in the PP. 
The WN is the client part of the channel, and the PP is the server part of the SSH channel. All communications between the PP and any services available on the WN side can be configured to use the SSH channel.
```

Once an SSH channel is established, a Port Forwarding layer, implemented on top of the Go SSH Lib, is used to secure any port to port communication initiated between the `Adaptor` and `Forwarder` and vice versa (see diagram). A client (1) seeking to communicate to the other side, comunicates with a local Inbound component (2) instead. The local Inbound act as a local server for the client. Once a connection is made to the Inbound, a tunnel (3) is created between the local Inbound and a remote Outbound component (4) via the SSH channel. The remote Outbound than acts as a client on behalf of the real client (1) and connect to the remote server (5) to establish an end to end session between the real client (1) and real server (5).

<p align="center">
  <img width="500"  src="ScPortForwarding.png">
</p>

> Once the SSH channel is created, clients at the Peer Pod VM side such as `KBC`, `Attestation Agent`, etc. may connect to services offered by the cluster side such as `KBS`, `Attestation Services`, etc. At the same time, clients located at the cluster side such as the `Runtime shim`, may connect to services offered by the Peer Pod VM side such as the `Kata Agent`.

Reviewer: ```suggestion
Once the SSH channel is created, clients at the Peer Pod side such as `Attestation Agent` connects to the services offered by the cluster side such as `KBS` using the channel. Likewise, clients located at the cluster side such as the `Kata shim`, connects to the services on the Peer Pod side such as the `Kata Agent` using the same channel.
```

> As shown in the following diagram, using `SecureComms`, each Peer Pod is consequently  served using a single communication channel implemented as an SSH channel - this single TCP session is initiated from the Worker Node and terminated at the Peer Pod and is being used for any control communication required between the Worker Node and Peer Pod and vice versa. This consolidation of Confidential Containers communication into a single TCP session improves the ability to control and secure the necessery communication to setup a Peer Pod. A firewall placed in front of the cluster Worker Node does not need to expose any cluster service externally. For example the communication between CDH and Trustee is done via a tunnel inside the SSH channel and does not require the Trustee service to be exposed to the internet.

Reviewer: ```suggestion
As shown in the following diagram, using `SecureComms`, each Peer Pod is consequently served using a single communication channel implemented as an SSH channel. This single TCP session is initiated from the Worker Node (WN) and terminated at the Peer Pod (PP). It is used for any control communication between the WN and PP and vice versa. This consolidation of communication between a WN and PP via a single SSH channel improves the operational aspects of WN and PP communication, especially when WN and PP are in different networks. The WN network does not need to open ingress ports to expose services to PP. For example, when the communication between CDH (in PP) and the Trustee (in the WN network) is via the SSH channel, the Trustee service port does not need to be open to enable communication with CDH (PP).
```

<p align="center">
  <img width="500"  src="SecureComms.png">
</p>

`SecureComms` in `Trustee Mode` uses the following sequence of steps:
- Worker Node creates keys for the Peer Pod and updates Trustee
- Worker Node creates the Peer Pod VM
- Worker Node establishes an "Attestation Phase" SSH channel with the Peer Pod VM
- Peer Pod VM attests and obtain keys from Trustee. The Peer Pod then signal to the Worker Node that the "Attestation Phase" has ended and terminates the "Attestation Phase" SSH channel.

> - Worker Node establishes an "Kubernetes Phase" SSH channel with the Peer Pod VM  - Both sides verify the identity of the other side using the keys delivered via Trustee.

Reviewer: ```suggestion
Worker Node establishes a "Kubernetes Phase" SSH channel with the Peer Pod VM - Both sides verify the identity of the other side using the keys delivered via Trustee.
```

`SecureComms` in `NoTrustee Mode` uses the following sequence of steps:
- Worker Node creates keys for the Peer Pod

> - Worker Node creates the Peer Pod VM and deliver the keys via daemonConfig as part of the userdata of the Cloudd API.

Reviewer: ```suggestion
Worker Node creates the Peer Pod VM and deliver the keys via daemonConfig as part of the userdata of the Cloud API.
```

- Peer Pod VM obtain keys from daemonConfig.

> - Worker Node establishes an "Kubernetes Phase" SSH channel with the Peer Pod VM  - Both sides verify the identity of the other side using the keys delivered via daemonConfig.

Reviewer: ```suggestion
- Worker Node establishes a "Kubernetes Phase" SSH channel with the Peer Pod VM - Both sides verify the identity of the other side using the keys delivered via daemonConfig.
```

> Once the "Kubernetes Phase" SSH channel is established, Secure Comms connects the `Runtime shim` to the `Kata Agent` and may also connect other services required by the Peer Pod VM or by containers running inside the Peer Pod.

Reviewer: nit-> Runtime shim -> Kata shim

Here, we describe how to set up the Secure Communications (`Secure Comms` for short) feature of the `cloud-api-adaptor` (`Adaptor`) and `agent-protocol-forwarder` (`Forwarder`).

The `Secure Comms` feature establishes an SSH channel by the `Adaptor`, as an SSH client, and the `Forwarder`, as an SSH server. All control communications then use this SSH channel between the `Adaptor` and `Forwarder`. This feature is an alternative to using `TLS encryption` feature between the `Adaptor` and `Forwarder`. 

> Both `Secure Comms` and `TLS encryption` supports a method of delivering keys to encrypt the control-plane traffic between the `Adaptor` and the `Forwarder` via user-data during pod VM creation. Delivering the keys via the user-data may allow a malicious entity on the cloud provider side to gain access to the keys and modify any communication between the WNs and PPs. For this reason, `Secure Comms` also supports an alternate mode where the keys are delivered to the PP via `Trustee`. The keys are delivered to the PP after an attestation phase (See diagram below). Delivering the keys via `Trustee` protects the PP from a malicious entity on the cloud provider side. 

Reviewer 2: Can we replace WN for worker node and PP for peer pods please to help with clarity?

Reviewer 2: IMHO this would also improve the readability of the code too.

<p align="center">
  <img width="500"  src="SecureComms.png">
</p>

`SecureComms` in `Trustee Mode` uses the following sequence of steps:
- Worker Node creates keys for the Peer Pod and updates Trustee
- Worker Node creates the Peer Pod VM
- Worker Node establishes an "Attestation Phase" SSH channel with the Peer Pod VM
- Peer Pod VM attests and obtain keys from Trustee. The Peer Pod then signal to the Worker Node that the "Attestation Phase" has ended and terminates the "Attestation Phase" SSH channel.
- Worker Node establishes a "Kubernetes Phase" SSH channel with the Peer Pod VM - Both sides verify the identity of the other side using the keys delivered via Trustee.

`SecureComms` in `NoTrustee Mode` uses the following sequence of steps:
- Worker Node creates keys for the Peer Pod

> Worker Node creates the Peer Pod VM and deliver the keys via daemonConfig as part of the userdata of the Cloud API.

Reviewer 2: You've missed the bullet point here

- Peer Pod VM obtain keys from daemonConfig.
- Worker Node establishes a "Kubernetes Phase" SSH channel with the Peer Pod VM - Both sides verify the identity of the other side using the keys delivered via daemonConfig.

Once the "Kubernetes Phase" SSH channel is established, Secure Comms connects the `Kata shim` to the `Kata Agent` and may also connect other services required by the Peer Pod VM or by containers running inside the Peer Pod.

See [Secure Comms Architecture Slides](./SecureComms.pdf) for more details.

## Setup for testing and for non-CoCo peerpods with NoTrustee mode

### Deploy CAA
Use any of the option for installing CAA depending on the cloud driver used.

### Activate Secure-Comms feature from CAA side
Activate Secure-Comms from CAA side by changing the `SECURE_COMMS` parameter of the `peer-pods-cm` configMap in the `confidential-containers-system` namespace to `"true"`.  

```sh
kubectl -n confidential-containers-system  get cm peer-pods-cm  -o yaml | sed "s/SECURE_COMMS: \"false\"/SECURE_COMMS: \"true\"/"|kubectl apply -f -

> ```

Reviewer 2: In this set-up `SECURE_COMMS_NO_TRUSTEE: true` is required, so can we add into into the  first section and not the optional cm edit?

Author: Done

### Activate Secure-Comms feature from CAA side
Activate Secure-Comms from CAA side by changing the `SECURE_COMMS` parameter of the `peer-pods-cm` configMap in the `confidential-containers-system` namespace to `"true"`.  

```sh
kubectl -n confidential-containers-system  get cm peer-pods-cm  -o yaml | sed "s/SECURE_COMMS: \"false\"/SECURE_COMMS: \"true\"/"|kubectl apply -f -
```

You may also include additional Inbounds and Outbounds configurations to the Adaptor side using the `SECURE_COMMS_INBOUNDS` and `SECURE_COMMS_OUTBOUNDS` config points.
You may also add Inbounds and Outbounds configurations to the Forwarder side using the `SECURE_COMMS_PP_INBOUNDS` and `SECURE_COMMS_PP_OUTBOUNDS` config points. [See more details regarding Inbounds and Outbounds below.](#adding-named-tunnels-to-the-ssh-channel)

Use `kubectl edit cm peer-pods-cm -n confidential-containers-system` to make such changes in the configMap, for example:
```sh
apiVersion: v1
data:
    ...
    SECURE_COMMS: "true"
    SECURE_COMMS_OUTBOUNDS: "KUBERNETES_PHASE:mytunnel:149.81.64.62:7777"
    SECURE_COMMS_PP_INBOUNDS: "KUBERNETES_PHASE:mytunnel:podns:6666"
    SECURE_COMMS_KBS_ADDR: "false"
    ...
```

> 

Reviewer: Some suggestions for the doc
1. Add a brief mention that this is an alternative to TLS encryption between the Adaptor and the Forwarder.
2. Add a mention about how the Attestation phase becomes optional if KBS is not used. Let's use the Attestation Phase (mandatory or optional) as a way to distinguish CoCo and Non-CoCo
3. Show two sequence of steps for SecureComms. One with Attestation Phase (and KBS). Another without the Attestation Phase. Use the same for setup instructions
4. If you can add a diagram showing a e2e example for inbound/outbound and namespaces from both the CAA and pod VM side, it'll be really helpful to visualise.

For example:
```sh
ExecStart=/usr/local/bin/agent-protocol-forwarder -kata-agent-namespace /run/netns/podns -secure-comms -secure-comms-inbounds KUBERNETES_PHASE:mytunnel:6666 -kata-agent-socket /run/kata-containers/agent.sock $TLS_OPTIONS $OPTIONS
```

Once you changed `podvm/files/etc/systemd/system/agent-protocol-forwarder.service`, you will need to [rebuild the podvm](./../podvm/README.md).

### Activate CAA Secure-Comms feature
Activate Secure-Comms of CAA by changing the `SECURE_COMMS` parameter of the `peer-pods-cm` configMap in the `confidential-containers-system` namespace to `"true"`.  

```sh
kubectl -n confidential-containers-system  get cm peer-pods-cm  -o yaml | sed "s/SECURE_COMMS: \"false\"/SECURE_COMMS: \"true\"/"|kubectl apply -f -
```

Set InitData to point KBC services to IP address 127.0.0.1 
```sh
cat <<EOF > /tmp/initdata.txt
algorithm = "sha384"
version = "0.1.0"

[data]
"aa.toml" = '''
[token_configs]
[token_configs.coco_as]
url = 'http://127.0.0.1:8080'

[token_configs.kbs]
url = 'http://127.0.0.1:8080'
'''
"cdh.toml"  = '''
socket = 'unix:///run/confidential-containers/cdh.sock'
credentials = []

[kbc]
name = 'cc_kbc'
url = 'http://127.0.0.1:8080'
'''

> EOF

Reviewer 2: I think we have the second issue here with repeated docs?

Author: Correct - thanks, fixed.
